File: scripts/sdk/transport_profile_probe.ts
import {
  KAMNClient,
  LiveTransportBackendAdapterError,
  LiveTransportKAMNClient,
  SDKError,
  TransportModeMismatchError,
} from "../../packages/kamn-sdk/src/index.ts";

function sanitize(value: string): string {
  return value.replaceAll("\n", " ");
}

function fail(message: string): never {
  console.log("status=error");
  console.log(`error=${sanitize(message)}`);
  process.exit(1);
}

try {
  const memory = new KAMNClient();
  const live = new LiveTransportKAMNClient("https://live.kamn.testnet/profile-probe-ts");

  let memoryExpected = "";
  let memoryFound = "";
  try {
    memory.assertTransportMode("live");
    fail("memory client unexpectedly accepted live mode assertion");
  } catch (error: unknown) {
    if (error instanceof TransportModeMismatchError) {
      memoryExpected = error.expected;
      memoryFound = error.found;
    } else if (error instanceof Error) {
      fail(error.message);
    } else {
      fail("unknown memory assertion error");
    }
  }

  let liveExpected = "";
  let liveFound = "";
  try {
    live.assertTransportMode("in-memory");
    fail("live client unexpectedly accepted in-memory mode assertion");
  } catch (error: unknown) {
    if (error instanceof TransportModeMismatchError) {
      liveExpected = error.expected;
      liveFound = error.found;
    } else if (error instanceof Error) {
      fail(error.message);
    } else {
      fail("unknown live assertion error");
    }
  }

  const successEndpoint = "https://live.kamn.testnet/profile-probe-ts-adapter";
  LiveTransportKAMNClient.registerBackendAdapter(successEndpoint, {
    invoke(request) {
      if (request.operation === "register") {
        return { status: "ok", value: "kamn:did:agent:backend-1" };
      }
      if (request.operation === "send") {
        return { status: "ok", value: "msg_backend_1" };
      }
      if (request.operation === "receive") {
        return {
          status: "ok",
          value: [
            {
              id: "msg_backend_1",
              from: "kamn:did:agent:backend-1",
              to: "kamn:did:agent:backend-2",
              body: "backend hello",
            },
          ],
        };
      }
      return { status: "ok", value: null };
    },
  });

  let backendAdapterRegisterId = "";
  let backendAdapterMessageId = "";
  let backendAdapterReceiveBody = "";
  try {
    const adapterClient = new LiveTransportKAMNClient(successEndpoint);
    backendAdapterRegisterId = adapterClient.register("autonomous", "claude-4", ["text"]);
    backendAdapterMessageId = adapterClient.send(
      "kamn:did:agent:backend-1",
      "kamn:did:agent:backend-2",
      "backend hello",
    );
    const received = adapterClient.receive("kamn:did:agent:backend-2");
    if (received.length !== 1) {
      fail("backend adapter receive returned unexpected message count");
    }
    backendAdapterReceiveBody = received[0].body;
  } finally {
    LiveTransportKAMNClient.clearBackendAdapters();
  }

  const failureEndpoint = "https://live.kamn.testnet/profile-probe-ts-adapter-fail";
  LiveTransportKAMNClient.registerBackendAdapter(failureEndpoint, {
    invoke(request) {
      if (request.operation === "register") {
        return { status: "ok", value: 7 };
      }
      if (request.operation === "send") {
        return { status: "error", reason: "backend_timeout" };
      }
      return { status: "error", reason: "policy_denied" };
    },
  });

  let backendAdapterInvalidResponseMessage = "";
  let backendAdapterErrorOperation = "";
  let backendAdapterErrorReason = "";
  let backendAdapterPolicyReason = "";
  try {
    const failingClient = new LiveTransportKAMNClient(failureEndpoint);
    try {
      failingClient.register("autonomous", "claude-4", ["text"]);
      fail("failing adapter unexpectedly accepted invalid register payload");
    } catch (error: unknown) {
      if (error instanceof SDKError || error instanceof Error) {
        backendAdapterInvalidResponseMessage = error.message;
      } else {
        fail("unknown adapter register failure");
      }
    }

    try {
      failingClient.send("kamn:did:agent:x", "kamn:did:agent:y", "hello");
      fail("failing adapter unexpectedly accepted send operation");
    } catch (error: unknown) {
      if (error instanceof LiveTransportBackendAdapterError) {
        backendAdapterErrorOperation = error.operation;
        backendAdapterErrorReason = error.reason;
      } else if (error instanceof Error) {
        fail(error.message);
      } else {
        fail("unknown adapter send failure");
      }
    }

    try {
      failingClient.receive("kamn:did:agent:y");
      fail("failing adapter unexpectedly accepted receive operation");
    } catch (error: unknown) {
      if (error instanceof LiveTransportBackendAdapterError) {
        backendAdapterPolicyReason = error.reason;
      } else if (error instanceof Error) {
        fail(error.message);
      } else {
        fail("unknown adapter receive failure");
      }
    }
  } finally {
    LiveTransportKAMNClient.clearBackendAdapters();
  }

  console.log("status=ok");
  console.log(`default_transport_mode=${memory.transportMode()}`);
  console.log(`live_transport_mode=${live.transportMode()}`);
  console.log(`memory_mismatch_expected=${memoryExpected}`);
  console.log(`memory_mismatch_found=${memoryFound}`);
  console.log(`live_mismatch_expected=${liveExpected}`);
  console.log(`live_mismatch_found=${liveFound}`);
  console.log(`backend_adapter_register_id=${backendAdapterRegisterId}`);
  console.log(`backend_adapter_message_id=${backendAdapterMessageId}`);
  console.log(`backend_adapter_receive_body=${backendAdapterReceiveBody}`);
  console.log(
    `backend_adapter_invalid_response_message=${sanitize(backendAdapterInvalidResponseMessage)}`,
  );
  console.log(`backend_adapter_error_operation=${backendAdapterErrorOperation}`);
  console.log(`backend_adapter_error_reason=${backendAdapterErrorReason}`);
  console.log(`backend_adapter_policy_reason=${backendAdapterPolicyReason}`);
} catch (error: unknown) {
  if (error instanceof SDKError || error instanceof Error) {
    fail(error.message);
  }
  fail("unknown error");
}

File: scripts/sdk/transport_profile_probe.py
#!/usr/bin/env python3
from __future__ import annotations

import sys
from pathlib import Path

ROOT_DIR = Path(__file__).resolve().parents[2]
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from kamn_sdk import (  # noqa: E402
    KAMNClient,
    LiveKAMNClient,
    LiveTransportBackendAdapterError,
    SDKError,
    TransportMode,
    TransportModeMismatchError,
)


def sanitize(value: str) -> str:
    return value.replace("\n", " ")


def main() -> int:
    try:
        memory = KAMNClient()
        live = LiveKAMNClient("https://live.kamn.testnet/profile-probe-python")

        try:
            memory.assert_transport_mode(TransportMode.LIVE)
        except TransportModeMismatchError as error:
            memory_expected = error.expected
            memory_found = error.found
        else:
            print("status=error")
            print("error=memory client unexpectedly accepted live mode assertion")
            return 1

        try:
            live.assert_transport_mode(TransportMode.IN_MEMORY)
        except TransportModeMismatchError as error:
            live_expected = error.expected
            live_found = error.found
        else:
            print("status=error")
            print("error=live client unexpectedly accepted in-memory mode assertion")
            return 1

        success_endpoint = "https://live.kamn.testnet/profile-probe-python-adapter"

        class SuccessAdapter:
            def invoke(self, request: dict[str, object]) -> dict[str, object]:
                operation = str(request.get("operation", ""))
                if operation == "register":
                    return {"status": "ok", "value": "kamn:did:agent:backend-1"}
                if operation == "send":
                    return {"status": "ok", "value": "msg_backend_1"}
                if operation == "receive":
                    return {
                        "status": "ok",
                        "value": [
                            {
                                "id": "msg_backend_1",
                                "from": "kamn:did:agent:backend-1",
                                "to": "kamn:did:agent:backend-2",
                                "body": "backend hello",
                            }
                        ],
                    }
                return {"status": "ok", "value": None}

        LiveKAMNClient.register_backend_adapter(success_endpoint, SuccessAdapter())
        try:
            adapter_client = LiveKAMNClient(success_endpoint)
            backend_register_id = adapter_client.register("autonomous", "claude-4", ["text"])
            backend_message_id = adapter_client.send(
                "kamn:did:agent:backend-1",
                "kamn:did:agent:backend-2",
                "backend hello",
            )
            backend_messages = adapter_client.receive("kamn:did:agent:backend-2")
            if len(backend_messages) != 1:
                print("status=error")
                print("error=backend adapter receive returned unexpected message count")
                return 1
            backend_receive_body = str(backend_messages[0].get("body", ""))
        finally:
            LiveKAMNClient.clear_backend_adapters()

        failure_endpoint = "https://live.kamn.testnet/profile-probe-python-adapter-fail"

        class FailureAdapter:
            def invoke(self, request: dict[str, object]) -> dict[str, object]:
                operation = str(request.get("operation", ""))
                if operation == "register":
                    return {"status": "ok", "value": 7}
                if operation == "send":
                    return {"status": "error", "reason": "backend_timeout"}
                return {"status": "error", "reason": "policy_denied"}

        LiveKAMNClient.register_backend_adapter(failure_endpoint, FailureAdapter())
        try:
            failing_client = LiveKAMNClient(failure_endpoint)
            try:
                failing_client.register("autonomous", "claude-4", ["text"])
                print("status=error")
                print("error=failing adapter unexpectedly accepted invalid register payload")
                return 1
            except SDKError as error:
                backend_invalid_response_message = str(error)

            try:
                failing_client.send("kamn:did:agent:x", "kamn:did:agent:y", "hello")
                print("status=error")
                print("error=failing adapter unexpectedly accepted send operation")
                return 1
            except LiveTransportBackendAdapterError as error:
                backend_error_operation = error.operation
                backend_error_reason = error.reason

            try:
                failing_client.receive("kamn:did:agent:y")
                print("status=error")
                print("error=failing adapter unexpectedly accepted receive operation")
                return 1
            except LiveTransportBackendAdapterError as error:
                backend_policy_reason = error.reason
        finally:
            LiveKAMNClient.clear_backend_adapters()

        print("status=ok")
        print(f"default_transport_mode={memory.transport_mode().value}")
        print(f"live_transport_mode={live.transport_mode().value}")
        print(f"memory_mismatch_expected={memory_expected}")
        print(f"memory_mismatch_found={memory_found}")
        print(f"live_mismatch_expected={live_expected}")
        print(f"live_mismatch_found={live_found}")
        print(f"backend_adapter_register_id={backend_register_id}")
        print(f"backend_adapter_message_id={backend_message_id}")
        print(f"backend_adapter_receive_body={backend_receive_body}")
        print(
            "backend_adapter_invalid_response_message="
            f"{sanitize(backend_invalid_response_message)}"
        )
        print(f"backend_adapter_error_operation={backend_error_operation}")
        print(f"backend_adapter_error_reason={backend_error_reason}")
        print(f"backend_adapter_policy_reason={backend_policy_reason}")
        return 0
    except SDKError as error:
        print("status=error")
        print(f"error={sanitize(str(error))}")
        return 1
    except Exception as error:  # pragma: no cover
        print("status=error")
        print(f"error={sanitize(str(error))}")
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
